Runtime/CsoundUnity.cs

```diff
@@ -250,6 +250,13 @@ public class CsoundUnity : MonoBehaviour
     /// If false, no processing occurs on the attached AudioClip
     /// </summary>
     [HideInInspector] public bool processClipAudio;
+	
+    /// <summary>
+    /// Sample Rate of the Csound instance
+    /// Used to calculate AudioSource pitch offset
+    /// Ensures that OnAudioFilterRead runs at/near the same rate as Csound
+    /// </summary>
+    [HideInInspector] public int sampleRate = 48000;

```

Comment:
Might be better to initialize this based on AudioSettings current sr.  Will push up change if this is desired 👍

Reply:
I agree. 
- I would add another field to enable the override, so that by default this field is hidden.
So:

`[HideInInspector] public bool overrideSampleRate;`

If this is true the sampleRate field is displayed in the inspector.

- Another change I would make, is that this field shouldn't be modified at runtime, so it needs to be disabled when the application is running, since the Csound sample rate is set only once when the instance is created, and changing it at runtime modifies the AudioSource's pitch value.

- The other change I would make is having this field as an enum with the available sample rates (the ones available for an AudioClip are 8000, 12025, 22050, 44100, 48000, 96000, 192000 hz) so that we can remove OnValidate.

- In my tests I see no improvement in the DSP consumption when running at lower sample rates, how do you measure it?

Reply:
I'll go ahead and make those changes

I was measuring the DSP CPU consumption in the Unity profiler.  What version of Unity are you running?  I wonder if there's a difference in how pitch is adjusted based on Unity version.  Here's a screenshot of the DSP consumption in profiler at the rates 48000, 22050, and 12025

![SampleRateDSPConsumption](https://user-images.githubusercontent.com/25762834/161403644-3f1600db-bd27-4426-b6dd-a2ca90eabe83.PNG)

Reply:
Hmm... so I installed Unity 2021.2.18f1, created a fresh project and then imported and updated CsoundUnity via the package manager but when I try and run the BasicTest sample file I get a crash to desktop.  This is without any of the changes I have made

Reply:
Okay so doing some testing with Csound 3.1.1, it seems that something changed so the pitch no longer seems to have any effect on the pitch of the csound output at all.  Going to do some digging and try to figure out why.  Good thing is it doesn't seem like it's anything on the Unity side that changed.  I see that there was an addition where a dummy clip is created to fix a spatialization issue, I'm wondering if maybe those fixes have something to do with it

Reply:
Alright so I narrowed it down, went through the commit history and it's definitely the fix for spatialization

https://github.com/rorywalsh/CsoundUnity/commit/712fdf6daf3929fd0544156a6ace8e887cb4ea54

Once I removed the uncommented portion on 1463 that was introduced and the spatialization fix, the functionality of the pitch slider returned to normal operation.  

What is the dummy audio clip for and what was the spatialization issue that was being encountered?

Reply:
@giovannibedetti Hey Giovanni.  Just following up on this.  What was the spatialization issue that Csound Unity had that your PR fixed?  Most just want to know what the issue was so I can see if I could keep the fix in tact but also find a workaround for how Unity handles its sample rate when a clip is present.

Reply:
I will try and see if I can find a way to preserve spatialization as well as the functionality of the pitch slider over the weekend

Reply:
Was this spatialization problem something that cropped up in a particular csounds and Unity version combination?  I have an older project where there doesn't seem to be any issues with spatialization so I'm assuming either a change on the Unity or the CsoundUnity side would have introduced it

Reply:
Gonna work on this tonight and see if I can figure out a workaround.  It is frustrating that OnAudioFilterRead doesn't seem to automatically match the sample rate of the dummy clip.  Looking into leveraging a PCMReaderCallback passed during the creation of the dummy clip, but some complained of high latency compared to OnAudioFilterRead so I'm skeptical.

Reply:
Already had some success leveraging a PCMReaderCallback.  No pitch modification is necessary with this approach, as it will read audio at the sample rate the dummy clip is created at.  Going to do some exploratory testing for bugs and log some performance comparisons and will get back to you.

Reply:
So utilizing the PCMReaderCallback with a streaming clip I was able to get spatialization and control of the sample rate, but unfortunately all performance benefits were lost with this approach due to how the PCMReaderCallback executes in comparison to OnAudioFilterRead.  It has less frequent peaks in DSP CPU consumption, but the peaks are much higher likely due to how the callback is being triggered (each call it likely has more samples to process at once compared to OnAudioFilterRead).  Of course, with this method none of the fixes for spatialization data are required, as it seems that the callback is pre-spatialization unlike OnAudioFilterRead.

I'm going to see if I can throw something together with the Unity native audio SDK.  It looks simple enough to route a signal to an AudioSource from OnAudioFilterRead in a way where spatialization data is preserved.  If it winds up taking too long or it consumes too much DSP CPU then I'll go ahead and close this PR out for the time being.

Reply:
Sorry for the (very long) delay on this one.  I spent some time experimenting with quite a few different solutions and then got sucked into a complete refactor of one of the systems I am working on that was unrelated to this issue.

So essentially, while the current spatialization solution preserves the 3d information, it seems to have a tremendous amount of zippering artifacts when objects start moving around at any reasonable rate of speed.  I'm not sure if it is outside the scope of this project (probably is) but would implementing a custom spatializer that incorporates a cross-fading buffer or something like that be desirable?  I'm assuming that the goal is to make this as usable as possible out the box, without the need for something like Microsoft's spatialization plugin for Unity.  

When I just went around the default spatializer and then used the pitch shift trick to slow the read rate of OnAudioFilterRead I gained all the performance benefits with none of the issues of the default spatializer.  This may all amount to "nothing to do here" but I figured I'd post anyways.  A less extreme solution than designing a custom spatialization solution would simply be adding a toggle to allow someone to bypass the parts of the code which locks the OnAudioFilterRead rate to the project settings sample rate.  This way you could toggle that so if you are using a third party or custom spatializer then OnAudioFilterRead rate isn't tied directly to the project's sample rate settings so users can gain the massive performance benefits of running Csound instances at lower sample rates.